### ISSUE TITLE
Implement ContractTemplate compiler

### ISSUE THREAD

**Issue by aslesarenko** [ergoplatform/sigmastate-interpreter#852]
The EIP-5 is created to standardize contract template format, which can be supported by ErgoTree compilers.

For ErgoScript compiler the syntax of contracts can be extended 

```scala
/** Description.
  * @param p1 
  * @param p2
  * @param p3
  * @return 
  */
@contract def MyContract(p1: Int, p2: GroupElement, p3: Coll[Byte]) = {
   // contract logic using p1 - p3
}
```

1) Some constants have names in the contract declaration, those names can be used to access the corresponding constant 
2) The mapping between constants and names is known during ErgoTree compilation and can be saved in the resulting ContractTemplate structure.
3) This name-index mapping can be made available and used during transaction construction by SDK libraries supporting EIP-5

**Comment by darkdrag00nv2** [comment]
@aslesarenko Can you please add more details on what needs to be done here? Like the proposed syntax etc.

I'll begin probing for this while we get the data structures work reviewed and merged.

**Comment by darkdrag00nv2** [comment]
This is up for grabs. I might not get time in the next few months for this.

**Comment by darkdrag00nv2** [comment]
Got some time now. Started taking a look.

**Comment by ross-weir** [comment]
I'll have an attempt at this.

@aslesarenko before I make a start could you please check my high-level understanding of what needs to be done:

1. Create a new `SigmaTemplateCompiler` that will take the contract template specified above and return a `ContractTemplate` instance
2. Create a new `SigmaTemplateParser` that will parse the docstring and extract parameter descriptions/etc as well as the function signature after `@contract`, build out the required `ContractTemplate` fields while parsing. When the parser gets to the opening `{` where the function body starts pass the rest of the contract on to `SigmaParser` that will produce the expression tree used for the `ContractTemplate`
3. At this point we should have a compiled `ContractTemplate` with all the required data

**Comment by aslesarenko** [comment]
Yes, looks like something I kept in mind. You can declare the necessary classes, put them on their places and let me review again.

**Comment by jozanek** [comment]
Was it maybe closed in #937?

**Comment by kushti** [comment]
> Was it maybe closed in [#937](https://github.com/ergoplatform/sigmastate-interpreter/pull/937)?

seems so, closing the issue
